src/dbnode/storage/limits/query_limits.go

```diff
@@ -0,0 +1,243 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package limits
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3/src/x/instrument"
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+)
+
+// DefaultLookback is the default lookback used for query limits.
+const DefaultLookback = time.Second * 15
+
+type queryLimits struct {
+	scope          tally.Scope
+	docsLimit      LookbackLimit
+	bytesReadLimit LookbackLimit
+}
+
+type lookbackLimit struct {
+	name    string
+	options LookbackLimitOptions
+	metrics lookbackLimitMetrics
+	recent  *atomic.Int64
+	stopCh  chan struct{}
+}
+
+type lookbackLimitMetrics struct {
+	recent     tally.Gauge
+	recentPeak tally.Gauge
+	total      tally.Counter
+	exceeded   tally.Counter
+}
+
+var (
+	_ QueryLimits   = (*queryLimits)(nil)
+	_ LookbackLimit = (*lookbackLimit)(nil)
+)
+
+// QueryLimits provides an interface for managing query limits.
+type QueryLimits interface {
+	DocsLimit() LookbackLimit
+	BytesReadLimit() LookbackLimit
+
+	AnyExceeded() error
+	Start()
+	Stop()
+}
+
+// LookbackLimit provides an interface for a specific query limit.
+type LookbackLimit interface {
+	Inc(new int) error
+	Exceeded() error
+	Start()
+	Stop()
+
+	// For testing purposes.
+	current() int64
+	reset()
+}
+
+// LookbackLimitOptions holds options for a lookback limit to be enforced.
+type LookbackLimitOptions struct {
+	// Limit past which errors will be returned.
+	Limit int64
+	// Lookback is the period over which the limit is enforced.
+	Lookback time.Duration
+}
+
+// NewQueryLimits returns a new query limits manager.
+func NewQueryLimits(
+	instrumentOpts instrument.Options,
```

Comment:
nit: what would you reckon about having `NewQueryLimits` do the validations of these options, then we can make the config parsing a little cleaner in that it doesn't need to validate options and just validates that the limit was created?